Tutorial for expect (#530)
usnistgov/metaschema: website/content/tutorials/3-constraints-expect/_index.md

---
title: "Complex Constraints: expect"
description: ""
---

# Complex Constraints

## Introduction

> In [the previous tutorial](/tutorials/2-constraints/), we refined a computer model and learned how to restrict or recommend preferred values for fields and flags with allowed-values constraints. But what do we if our use case requires values to have a consistent style or structure, but there is no predetermined list of values our stakeholders know upfront? Additionally, what do we do if we have such requirements for field and flag values, but they must also correlate to related field or flag values in the same document instance, or maybe multiple related instances the software processes at the same time?

Reviewer: "Additionally, what do we do if we have such requirements for field and flag values" What does "such" reference?

Author: Such is short for requirements where "there is no predetermined list of values our stakeholders know upfront" from the previous sentence.

Reviewer: I don't think it is clear in the text what "such" references. Can you adjust the text to make this more clear?

Author: Sure thing.

Author: I tried to handle this and some other final cleanup in https://github.com/usnistgov/metaschema/pull/530/commits/7719523f084f9dd055f2676d6a86881f9f77e67e.

> To precisely control the structure of values and their relationship to one another without a predetermined list of values, a model developer can use `expect` and `matches` constraints, which we will examine the tutorial below.

Reviewer: ```suggestion
To precisely control the structure of values and their relationship to one another without a predetermined list of values, a model developer can use the `expect` and `matches` constraints, which we will examine the tutorial below.
```

    type: atx
    cpu:
      vendor:
        id: vendor2
        name: ISA Corp
        address: 2000 K Street NW Washington, DC 20002
        website: https://example.com/isacorp/
      architecture: arm64
      product-name: Superchip Model 1 4-core Processor
      speed: 4.7 gigahertz
    ata-socket:
      vendor:
        id: vendor2
        name: ISA Corp
        address: 2000 K Street NW Washington, DC 20002
        website: https://example.com/isacorp/
      product-name: AwesomeSAS Model 1 Storage Socket
      type: sas
    memory-modules:
    - vendor:
        id: vendor3
        name: Massive Memory Corp
        address: 3000 K Street NW Washington, DC 20003
        website: https://example.com/massive-memory-corp/
      product-name: Model 3 DDR4-3200 8GB (Module 1)
      byte-size: 8589934592
    - byte-size: 8589934592
      product-name: Model 3 DDR4-3200 8GB (Module 2)
      vendor:
        address: 3000 K Street NW Washington, DC 20003
        id: vendor3
        name: Massive Memory Corp
        website: https://example.com/massive-memory-corp/
{{< /highlight >}}
{{% /tab %}}
{{< /tabs >}}

## Logically testing values with `expect`

> Given previous success with our recent rollout of new constraints, they provide us a new requirement. It is highly unusual for computer memory, regardless of form factor, to not be an even number divisible by two. Furthermore, per requirements from their computer manufacturing consortium and the information model you help maintain, best practice recommends that memory modules must be increments of 1 megabyte (1024 bytes) and all memory modules should be the same size in bytes if more than one memory module is present. Theoretically, we could amend the model to enumerate all possible numeric bytes values that could meet the first requirement, but the secondary best practice requirements cannot be expressed with `allowed-values`. Even so, we would make a very large model that performs very inefficiently if we considered it for the first requirement. However, We can use an `expect` constraint to meet all these requirements.

Reviewer: ```suggestion
Given previous success with our recent rollout of new constraints, we have a new requirement. It is highly unusual for computer memory, regardless of form factor, to not be an even number divisible by two. Furthermore, per requirements from their computer manufacturing consortium and the information model you help maintain, best practice recommends that:

 - memory modules must be increments of 1 megabyte (1024 bytes), and
 - all memory modules should be the same size in bytes if more than one memory module is present.
 
Theoretically, we could amend the model to enumerate all possible numeric bytes values that could meet the first requirement, but the second best practice requirement cannot be expressed with `allowed-values`. Even so, we would make a very large model that performs very inefficiently if we considered enumerating all possible values for the first requirement. We can use an `expect` constraint to meet all these requirements in a much more efficient and manageable way.
```

                </model>
              </define-assembly>
              <define-field name="product-name" as-type="string" min-occurs="1" max-occurs="1">
                <formal-name>Product Name</formal-name>
                <description>The product name from the vendor of the computer part.</description>
              </define-field>
              <define-field name="type" as-type="string" min-occurs="1" max-occurs="1">
                <formal-name>Expansion Card Type</formal-name>
                <description>The type of expansion card on a motherboard of a computer, such as <code>pci</code> (PCI, e.g. Peripheral Component Interconnect), <code>pcie</code> (PCI Express), or an alternative.</description>
              </define-field>
            </model>
          </define-assembly>
        </model>
        <constraint>
          <expect id="memory-same-byte-size" level="WARNING" target="." test="(sum(./memory/byte-size) mod ./memory/byte-size[1]) = 0">
            <message>All memory modules SHOULD be the same byte size for a computer.</message>
          </expect>          
        </constraint>
      </define-assembly>
    </model>
  </define-assembly>
</METASCHEMA>
```

> The three new `expect` constraints define one mandatory requirement and two optional best practice requirements without exhaustively defining every possible value. Each constraint has a unique `id` for us tool developers to identify the constraints during tool development and potentially use as inputs and outputs for Metaschema processor software itself. These constraints use the flexibility of Metapath to use different context focuses to make the `test` attribute more concise or intuitive for fellow developers.

Reviewer: ```suggestion
The three new `expect` constraints define one mandatory requirement and two optional best practice requirements without exhaustively defining every possible value. Each constraint has a unique `id` that helps tool developers to identify the constraints during tool development and to potentially use as inputs and outputs for Metaschema processor software performing validation using these constraints. These constraints use the flexibility of Metapath to use different context focuses to make the `test` attribute more concise or intuitive for fellow developers.
```

> The `level` attribute defines the impact and potential processing behavior for Metaschema processors during the analysis of document instances. The `level` of a constraint can be `INFORMATIONAL`, `WARNING`, `ERROR`, or `CRITICAL`. Given the new requirements from the stakeholder, failing the `memory-divisible-two` constraint indicates invalid data and necessarily an error condition, so it is an `ERROR`. The `memory-divisible-megabyte` and `memory-same-byte-size` do not encode mandatory requirements, but rather best practices. Therefore, their `level` is `WARNING`.

Reviewer: ```suggestion
The `level` attribute defines the impact and potential processing behavior for Metaschema processors during the analysis of document instances. The `level` of a constraint can be `INFORMATIONAL`, `WARNING`, `ERROR`, or `CRITICAL`. Given the new requirements we are implementing, failing the `memory-divisible-two` constraint indicates invalid data, requiring an error condition, so we mark the level as an `ERROR`. The `memory-divisible-megabyte` and `memory-same-byte-size` constraints do not encode mandatory requirements, but rather best practices. Therefore, their `level` is `WARNING`.
```

Our Metaschema processors must evaluate the `test` Metapath against the current context focus from evaluating the `target` (using the `.` operator in Metapath). That result becomes the current context focus in the `test` (the  value of `.` in this expression), and the processor must evaluate the Metapath expression and verify the result is boolean true or false.

> The table below identifies the constraints and evaluations of the new constraints in the updated model against the previous sample document instances. Because we are able to control the `target` and `test` context relative to their definitions in the model's assemblies, fields, and flags, we process each result individually for the first two constraints, but process the sequence of all matches for the last constraint all at once. For the latter, we use the more advanced Metapath syntax in the `test` to query each matching assembly of the sequence, filter to only the specific fields' value , and sum them. The data in the table is truncated, but the processor will use process the full `memory` assemblies as the resulting match, and subsequently the `test` Metapath filters on the values of the `byte-size` of each dynamically to compute the sum, then perform the necessary modulo division by the first result (there must always be at least one, never zero, per our module `min-occurs="1"`).

Reviewer: ```suggestion
The table below identifies the constraints and resulting evaluations based on the updated model using the previous sample document instances. Because we are able to control the `target` and `test` context relative to their definitions in the model's assemblies, fields, and flags, we process each result individually for the first two constraints, but process the sequence of all matches for the last constraint all at once. For the latter, we use more advanced Metapath syntax in the `test` to query each matching assembly of the sequence, filter to only the specific fields' value , and sum them. The data in the table is truncated, but the processor will process all the `memory` assemblies as the resulting match, and subsequently the `test` Metapath filters on the values of the `byte-size` of each dynamically to compute the sum, then perform the necessary modulo division by the first result (there must always be at least one, never zero, per our module `min-occurs="1"`).
```

The last sentences are a bit awkward. Consider revising them.

Author: Don't I know it. I will think about this then and clean them up instead of just accepting this recommendation. (I am presuming this is what you mean, and I can totally accept that criticism.)

Author: Well maybe I will just pull in the batch and then word smith it tomorrow morning, but either way I agree it is awkward.

{{% /tab %}}
{{< /tabs >}}

We can use the samples above to simulate a scenario where a developer or their tool define an invalid memory module configuration. The table below identifies constraints and their evaluation against the new sample. We can observe the processor should report on specific constraint errors for some fields, but not others.

| Constraint ID             | Evaluated Target                                                  | Evaluated Test            | Result |
|---------------------------|-------------------------------------------------------------------|---------------------------|--------|
| memory-divisible-two      | `<!-- 1st memory module -->`<br/>`<byte-size>8589934591</byte-size>` |  `(8589934591 mod 2) = 0`    | false  |
|                           | `<!-- 2nd memory module -->`<br/>`<byte-size>8589934592</byte-size>` |  `(8589934592 mod 2) = 0`    | true   |
| memory-divisible-megabyte | `<!-- 1st memory module -->`<br/>`<byte-size>8589934591</byte-size>` |  `(8589934591 mod 1024) = 0` | false  |
|                           | `<!-- 2nd memory module -->`<br/>`<byte-size>8589934592</byte-size>` |  `(8589934592 mod 1024) = 0` | true   |

> | memory-same-byte-size     | Sequence of <br/> (`<memory>...<product-name>... (Module 1)</product-name><byte-size>8589934591</byte-size></memory>`, <br/>`<memory>...<product-name>... (Module 2)</product-name><byte-size>8589934592</byte-size></memory>`) | `(sum((8589934592, 8589934592)) mod 8589934592) = 0)` |  true |

Reviewer: Is there a reason this table is duplicated from above?

Author: If you look some the values are changing from the math and the results column accordingly switches from true to false. I just wanted to start explaining how a developer could conceptualize "what is the Metapath doing and why?" match by match and eval by eval.

## Deprecation warnings with `expect`

> With our use of the `expect` constraint for logical value testing, our stakeholders were very pleased we can develop and test small model enhancements incrementally and quickly. However, the adoption of a standard information model and data formats have surfaced some problems and challenges they were unaware of, so they need our help. An important group in their consortium identified a risk around using `byte-size` with binary base-two number for memory or storage capacity. The use of [binary prefixes and units](https://en.wikipedia.org/wiki/Binary_prefix) are ambiguous in the new model. Some vendors use binary base-two sizes or decimal base-ten sizes, but this field in the model makes tooling inconsistent and document instances unclear. A majority in the consortium decided updates to documentation (sourced from our Metaschema modules) are not enough. There must be a way to identify size, its unit prefix (binary or decimal), and optionally a unit so developers and software do not need to compute the full byte size. Our stakeholders are very concerned because the consortium also wants to allow the old approach in our information model and mark this method as deprecated. The intent will be to remove after twelve to twenty four months.

Reviewer: ```suggestion
With our use of the `expect` constraint for logical value testing, our stakeholders were very pleased we can develop and test small model enhancements incrementally and quickly. However, the adoption of a standard information model and data formats have surfaced some problems and challenges they were unaware of, so they need our help. An important group in their consortium identified a risk around using `byte-size` with binary base-two number for memory or storage capacity. The use of [binary prefixes and units](https://en.wikipedia.org/wiki/Binary_prefix) are ambiguous in the new model. Some vendors use binary base-two sizes or decimal base-ten sizes, but this field in the model makes tooling inconsistent and document instances unclear. A majority in the consortium decided updates to documentation (sourced from our Metaschema modules) are not enough. There must be a way to identify size, its unit prefix (binary or decimal), and optionally a unit so developers and software do not need to compute the full byte size. Our stakeholders are very concerned because the consortium also wants to allow the old approach in our information model and mark this method as deprecated. The intent will be to remove the deprecated approach after twelve to twenty four months.
```

                <formal-name>Product Name</formal-name>
                <description>The product name from the vendor of the computer part.</description>
              </define-field>
              <define-field name="type" as-type="string" min-occurs="1" max-occurs="1">
                <formal-name>Expansion Card Type</formal-name>
                <description>The type of expansion card on a motherboard of a computer, such as <code>pci</code> (PCI, e.g. Peripheral Component Interconnect), <code>pcie</code> (PCI Express), or an alternative.</description>
              </define-field>
            </model>
          </define-assembly>
        </model>
        <constraint>
          <expect id="memory-same-byte-size" level="CRITICAL" target="." test="if (count(./memory/byte-size) > 0) then (sum(./memory/byte-size) mod ./memory/byte-size[1]) = 0 else (sum(./memory/size) mod ./memory/size[1]) = 0">
              <message>All memory modules SHOULD be the same size or byte-size for a computer.</message>
          </expect>
        </constraint>
      </define-assembly>
    </model>
  </define-assembly>
</METASCHEMA>
```

> With our modifications to the model with the `choice` construct, we enabled our Metaschema-enabled tooling to support a field of `byte-size` or `size` where previously only the former was valid. For `size`, we added important constraints to its flag combinations to identify which unit (bits or bytes), unit prefix, unit system, and base (digital or binary) of the size's count. With the `byte-size` field, we add an `expect` constraint with a specific message to only present warning if information if model and tool developers use this deprecated field in document instances. Otherwise it can be ignored. Moreover, we refactored the previous size count constraints to check the logical constraint whether the developers use the deprecated field or the replacement field with the same logic using an `if` `else` expression.

Reviewer: ```suggestion
With our modifications to the model with the `choice` construct, we enabled our Metaschema-enabled tooling to support a field of `byte-size` or `size` where previously only the former was valid. For `size`, we added important constraints to its flag combinations to identify which unit (bits or bytes), unit prefix, unit system, and base (digital or binary) of the size's count. With the `byte-size` field, we add an `expect` constraint with a specific message to only present a warning if developers use this deprecated field in document instances; otherwise, the field can be ignored. Moreover, we refactored the previous size count constraints to check the logical constraint regardless of when the deprecated field or the replacement field is used. We support this with the same logic using an `if` `else` expression.
```

              </define-field>
            </model>
          </define-assembly>
        </model>
        <constraint>
          <expect id="memory-same-byte-size" level="WARNING" target="." test="(sum(./memory/byte-size) mod ./memory/byte-size[1]) = 0">
            <message>All memory modules SHOULD be the same byte size for a computer.</message>
          </expect>          
        </constraint>
      </define-assembly>
    </model>
  </define-assembly>
</METASCHEMA>
```

The three new `expect` constraints define one mandatory requirement and two optional best practice requirements without exhaustively defining every possible value. Each constraint has a unique `id` that helps tool developers to identify the constraints during tool development and to potentially use as inputs and outputs for Metaschema processor software performing validation using these constraints. These constraints use the flexibility of Metapath to use different context focuses to make the `test` attribute more concise or intuitive for fellow developers.

> The `level` attribute defines the impact and potential processing behavior for Metaschema processors during the analysis of document instances. The `level` of a constraint can be `INFORMATIONAL`, `WARNING`, `ERROR`, `CRITICAL`, or `DEBUG`. Given the new requirements we are implementing, failing the `memory-divisible-two` constraint indicates invalid data, requiring an error condition, so we mark the level as an `ERROR`. The `memory-divisible-megabyte` and `memory-same-byte-size` constraints do not encode mandatory requirements, but rather best practices. Therefore, their `level` is `WARNING`.

Reviewer: ```suggestion
The `level` attribute defines the impact and potential processing behavior for Metaschema processors during the analysis of document instances. The `level` of a constraint can be `DEBUG`, `INFORMATIONAL`, `WARNING`, `ERROR`, or `CRITICAL`. Given the new requirements we are implementing, failing the `memory-divisible-two` constraint indicates invalid data, requiring an error condition, so we mark the level as an `ERROR`. The `memory-divisible-megabyte` and `memory-same-byte-size` constraints do not encode mandatory requirements, but rather best practices. Therefore, their `level` is `WARNING`.
```
